cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AbstractAppLifecycleHttpHandler.java

```diff
@@ -0,0 +1,128 @@
+import io.cdap.cdap.gateway.handlers.util.AbstractAppFabricHttpHandler;
+import io.cdap.cdap.internal.app.deploy.ProgramTerminator;
+import io.cdap.cdap.internal.app.deploy.pipeline.ApplicationWithPrograms;
+import io.cdap.cdap.proto.ApplicationRecord;
+import io.cdap.cdap.proto.id.ApplicationId;
+import io.cdap.cdap.proto.id.EntityId;
+import io.cdap.cdap.proto.id.KerberosPrincipalId;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.proto.id.ProgramId;
+import java.util.Optional;
+import javax.annotation.Nullable;
+
+public abstract class AbstractAppLifecycleHttpHandler extends AbstractAppFabricHttpHandler {
+  protected final CConfiguration configuration;
+  protected final NamespaceQueryAdmin namespaceQueryAdmin;
+  protected final ProgramRuntimeService runtimeService;
+
+  public AbstractAppLifecycleHttpHandler(CConfiguration configuration,
+      NamespaceQueryAdmin namespaceQueryAdmin,
+      ProgramRuntimeService runtimeService) {
+    this.configuration = configuration;
+    this.namespaceQueryAdmin = namespaceQueryAdmin;
+    this.runtimeService = runtimeService;
+  }
+
+  protected ApplicationId validateApplicationVersionId(NamespaceId namespaceId, String appId,
+      String versionId) throws BadRequestException {
+    if (appId == null) {
+      throw new BadRequestException("Path parameter app-id cannot be empty");
+    }
+    if (!EntityId.isValidId(appId)) {
+      throw new BadRequestException(String.format("Invalid app name '%s'", appId));
+    }
+    if (versionId == null) {
+      throw new BadRequestException("Path parameter version-id cannot be empty");
+    }
+    if (EntityId.isValidVersionId(versionId)) {
+      return namespaceId.app(appId, versionId);
+    }
+    throw new BadRequestException(String.format("Invalid version '%s'", versionId));
+  }
+
+  protected NamespaceId validateNamespace(@Nullable String namespace)
+      throws BadRequestException, NamespaceNotFoundException, AccessException {
+
+    if (namespace == null) {
+      throw new BadRequestException("Path parameter namespace-id cannot be empty");
+    }
+
+    NamespaceId namespaceId;
+    try {
+      namespaceId = new NamespaceId(namespace);
+    } catch (IllegalArgumentException e) {
+      throw new BadRequestException(String.format("Invalid namespace '%s'", namespace), e);
+    }
+
+    try {
+      if (!namespaceId.equals(NamespaceId.SYSTEM)) {
+        namespaceQueryAdmin.get(namespaceId);
+      }
+    } catch (NamespaceNotFoundException | AccessException e) {
+      throw e;
+    } catch (Exception e) {
+      // This can only happen when NamespaceAdmin uses HTTP calls to interact with namespaces.
+      // In AppFabricServer, NamespaceAdmin is bound to DefaultNamespaceAdmin, which interacts directly with the MDS.
+      // Hence, this exception will never be thrown
+      throw Throwables.propagate(e);
+    }
+    return namespaceId;
+  }
+
+  protected ProgramTerminator createProgramTerminator() {
```

Comment:
It was already in the `AppLifecycleHttpHandler` and used in multiple places. Here we have only moved it up (to the abstract class, so that it can be used in the internal handler as well).